book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,121 @@
+
+Define the function `is_locked(a, b)` to be: `a + a.lockout <= b`
+
+# Determining Vote Safety From a Snapshot and Locktower
+
+Define the "safety" condition to be: 
+
+Given a `BankForks`, a validator can run some procedure to determine whether it can vote on a frozen bank `T` (frozen banks must be present in `BankForks.banks`), without violating the lockouts of any vote in `L`. The procedure for this is:
+
+1) If `T` < last locktower vote, return false,
+2) Apply `T` to locktower state `L`. Pop off all votes `S_i` where `!is_locked(S_i, T)`. `T` must now be the top of the tower.
+3) If for any remaining vote `L_i` in `L` `!is_ancestor(L_i, T)`, return false.
+4) Otherwise, return true.
+
+# Achieving Safety
+
+Define the "Safety Criteria" to be: Given a `BankForks` any descendant of `S`, `S_d` that has a bank `B_d`  that is present in `BankForks.banks`, and any slot `L_i` in locktower, the validator is able to determine `is_ancestor(L_i, S_d)`.
+
+Assume the "Safety Criteria" is true, we show we can then achieve the "safety" condition:
+
+Proof: 
+A validator wants to determine whether it can vote on `T`. `T` must be a descendant of `S` because the validator does not play any state for non-descendants of `S` when booting from the snapshot. From assuming "Safety Criteria" above, this means for any `L_i` we can determine `is_ancestor(L_i, T)`. This means we have all the tools to run the algorithm from the safety definition.
+
+Thus to achieve safety, we want to design the snapshotting system such that the "Safety Criteria" is met.
+
+# Implementing "Safety Criteria"
+
+"Range Assumption": In our implementation, we assume that at the time the validator starts from the snapshot, `L_i + N >= S` for some large `N`. 
+
+Snapshot Design:
+
+The snapshot is augmented to store the last `N` ancestors of `S`. These ancestors are  incorporated into the bank hash so they can be verified when a validator unpacks a snapshot. Call this set `N_Ancestors`.
+
+Boot Procedure: 
+
+1) We reject any snapshots `S` where `S` is less than the root of `L` and `S` is not in the list of roots in blocktree because this means `S` is for a different fork than the one this validiator last rooted. Thus it's critical for consistency in `replay_stage` that the order of events when setting a new root is:
+    a) Write the root to locktower
+    b) Write the root to blocktree
+    c) Generate snapshot for that root
+
+2) On startup, make sure the current root in locktower exists in blocktree, if not (there was a crash between 2a and 2b), then rewrite the root to blocktree.
+
+3) On startup, the validator boots from the snapshot `S`, then replays all descendant blocks of `S` that exist in this validator's ledger, building banks which are then stored in an output `BankForks`. This is done in `blocktree_processor.rs`. The root of this `BankForks` is set to `S`.
+
+
+We now implement the "Safety Criteria" in cases:
+
+Case 1: Calculating `is_ancestor(L_i, S_d)` when `L_i` < `S`:
```

Comment:
We also have the case of `L_i` == `S`. That might fit is by changing Case 1 to `L_i` <= `S`

Reply:
Yup folded it into Case 2!